sinabs/backend/dynapcnn/nir_graph_extractor.py

```diff
@@ -0,0 +1,501 @@
+        dcnnl_map = collect_dynapcnn_layer_info(
+            indx_2_module_map = self.indx_2_module_map,
+            edges = self.edges,
+            nodes_io_shapes=self.nodes_io_shapes,
+            entry_nodes=self.entry_nodes,
+        )
+
+        # build `DynapcnnLayer` instances from mapper.
+        dynapcnn_layers, destination_map, entry_points = (
+            construct_dynapcnnlayers_from_mapper(
+                dcnnl_map=dcnnl_map,
+                discretize=discretize,
+                rescale_fn=weight_rescaling_fn,
+            )
+        )
+
+        # Instantiate the DynapcnnNetworkModule
+        return DynapcnnNetworkModule(
+            dynapcnn_layers, destination_map, entry_points
+        )
+
+    def remove_nodes_by_class(self, node_classes: Tuple[Type]):
+        """Remove nodes of given classes from graph in place.
+
+        Create a new set of edges, considering layers that `DynapcnnNetwork` will ignore. This
+        is done by setting the source (target) node of an edge where the source (target) node
+        will be dropped as the node that originally targeted this node to be dropped.
+
+        Will change internal attributes `self._edges`, `self._entry_nodes`,
+        `self._name_2_indx_map`, and `self._nodes_io_shapes` to reflect the changes.
+
+        Parameters
+        ----------
+        - node_classes (tuple of types):
+            Layer classes that should be removed from the graph.
+
+        """
+        # Compose new graph by creating a dict with all remaining node IDs as keys and set of target node IDs as values
+        source2target: Dict[int, Set[int]] = {
+            node: self._find_valid_targets(node, node_classes)
+            for node in self.sorted_nodes
+            # Skip nodes that are to be removed
+            if not isinstance(self.indx_2_module_map[node], node_classes)
+        }
+
+        # remapping nodes indices contiguously starting from 0
+        remapped_nodes = {
+            old_idx: new_idx
+            for new_idx, old_idx in enumerate(sorted(source2target.keys()))
+        }
+
+        # Parse new set of edges based on remapped node IDs
+        self._edges = {
+            (remapped_nodes[src], remapped_nodes[tgt])
+            for src, targets in source2target.items()
+            for tgt in targets
+        }
+
+        # Update internal graph representation according to changes
+        self._update_internal_representation(remapped_nodes)
+
+    def get_node_io_shapes(self, node: int) -> Tuple[torch.Size, torch.Size]:
+        """Returns the I/O tensors' shapes of `node`.
+
+        Returns
+        ----------
+        - input shape (torch.Size): shape of the input tensor to `node`.
+        - output shape (torch.Size): shape of the output tensor from `node`.
+        """
+        return (
+            self._nodes_io_shapes[node]["input"],
+            self._nodes_io_shapes[node]["output"],
+        )
+
+    def verify_graph_integrity(self):
```

Comment:
This should also verify if that are self-recurrent nodes - i.e., edges in the format (X, X).